src/ddmd/root/filename.d

```diff
@@ -721,3 +724,111 @@ nothrow:
         return str;
     }
 }
+
+version(Windows)
+{
+    /*
+      The code before used the POSIX function `mkdir` on Windows. That
+      function is now deprecated and fails with long paths, so instead
+      we use the newer `CreateDirectoryW`.
+
+      `CreateDirectoryW` is the unicode version of the generic macro
+      `CreateDirectory`.  `CreateDirectoryA` has a file path
+      limitation of 248 characters, `mkdir` fails with less and might
+      fail due to the number of consecutive `..`s in the
+      path. `CreateDirectoryW` also normally has a 248 character
+      limit, unless the path is absolute and starts with `\\?\`. Note
+      that this is different from starting with the almost identical
+      `\\?`.
+
+      Please consult
+      https://msdn.microsoft.com/en-us/library/windows/desktop/aa363855(v=vs.85).aspx
+    */
+    private int _mkdir(const(char)* path) nothrow
+    {
+        import core.sys.windows.winerror: ERROR_ALREADY_EXISTS;
+        import core.stdc.errno: errno, EEXIST;
+
+        const extendedPath = path.toExtendedLengthPath();
+
+        // the CreateDirectoryW function takes a pointer to a structure,
+        // but since we're not interested in using it it's declared void*
+        // and we pass null for the default behaviour.
+        void* securityAttributes;
+        const createRet = () @trusted
+        {
+            return CreateDirectoryW(&extendedPath[0], securityAttributes);
+        }();
+        const lastError = GetLastError();
+
+        // Preserve compatibility with mkdir since the calling code expects
+        // errno to be set.
+        if (createRet == 0 && lastError == ERROR_ALREADY_EXISTS)
+            errno(EEXIST);
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
+    }
+
+    // Converts an UTF8 null-terminated string to an array of wchar that's null
+    // terminated so it can be passed to Win32 APIs.
+    private wchar[] toWStringz(const(char*) str) pure nothrow {
```

Comment:
Actually, on further reflection, what I'd be tempted to do would be to implement `toExtendedLengthPath` as a function that took a `wchar[]` as a buffer and then pass it a slice of a large static array and try to implement that such that it didn't allocate anything (particularly in light of the fact that dmd is kind of sucky with regards to memory usage), but I'm not sure that that's worth it, and it would max out the path length at whatever the static array's length was, which could be large, but it would still limit it more than would be the case with a dynamic array.

But if the dmd devs have no problem with constructing the dynamic array like this, then this should be fine.

Reply:
The problem is, as you stated, picking that size. It completely depends on what happens at runtime. The best solution would probably be to use a static array and allocate if it needs more memory. I've changed it to be a little bit more complicated but faster for usual workloads.